components/anemometer/anemometer.c

```diff
@@ -0,0 +1,137 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file anemometer.c
+ *
+ * ESP-IDF driver for impulse wind speed sensors(anemometers)
+ *
+ * Ported from esp-open-rtos
+ *
+ * Copyright (c) 2024 Jakub Turek <qb4.dev@gmail.com>
+ *
+ * BSD Licensed as described in the file LICENSE
+ */
+#include "anemometer.h"
+
+#include <freertos/FreeRTOS.h>
+#include <freertos/task.h>
+#include <esp_timer.h>
+#include <driver/gpio.h>
+#include <esp_idf_lib_helpers.h>
+
+#if HELPER_TARGET_IS_ESP32
+static portMUX_TYPE mux = portMUX_INITIALIZER_UNLOCKED;
+#define PORT_ENTER_CRITICAL() portENTER_CRITICAL(&mux)
+#define PORT_EXIT_CRITICAL() portEXIT_CRITICAL(&mux)
+
+#elif HELPER_TARGET_IS_ESP8266
+#define PORT_ENTER_CRITICAL() portENTER_CRITICAL()
+#define PORT_EXIT_CRITICAL() portEXIT_CRITICAL()
+
+#else
+#error cannot identify the target
+#endif
+
+#define CHECK_ARG(VAL) do { if (!(VAL)) return ESP_ERR_INVALID_ARG; } while (0)
+
+typedef struct {
+	gpio_num_t input_pin;      //!< GPIO input pin
+	float sf;                  //!< scale factor
+	uint32_t pps;              //!< measured pulses count per second
+	TickType_t init_tick;      //!< measurement init tick
+} anemometer_priv_t;
+
+static inline void pps_iterate(anemometer_priv_t *priv, TickType_t ticks)
+{
+    if(ticks - priv->init_tick > pdMS_TO_TICKS(1000)){
+        priv->init_tick = ticks;
+        priv->pps = 0;
+    }
+}
+
+static void IRAM_ATTR gpio_isr_handler(void* arg)
+{
+    anemometer_priv_t *priv = (anemometer_priv_t *)arg;
+    TickType_t ticks = xTaskGetTickCountFromISR();
+    pps_iterate(arg,ticks);
+    priv->pps++;
+}
+
+anemometer_t anemometer_init(const anemometer_config_t *conf)
+{
+	gpio_config_t io_conf;
+    anemometer_priv_t *priv;
+
+    priv = (anemometer_priv_t *)calloc(1,sizeof(anemometer_priv_t));
+    if(priv == NULL){
+        return NULL;
+    }
+
+    priv->input_pin = conf->input_pin;
+    priv->sf = conf->scale_factor ? conf->scale_factor : ANEMOMETER_DEFAULT_SF;
+
+    /* setup GPIO */
+    io_conf.intr_type = GPIO_INTR_POSEDGE;
+    io_conf.pin_bit_mask = (1 << priv->input_pin);
+    io_conf.mode = GPIO_MODE_INPUT;
+    io_conf.pull_up_en = GPIO_PULLUP_ENABLE;
+    gpio_config(&io_conf);
+
+    /* enable interrupts */
+    gpio_install_isr_service(0);
```

Comment:
What if the isr service is already installed in another component?

Reply:
added check like in wiegand driver